File: ADTs/CPP-DS/AdapterExamples.cpp
#include <iostream> 
#include <algorithm> 
#include <array> 
#include <iterator> 
#include <string> 
#include <vector>
#include <list>
#include <stack>
#include <queue>
#include <deque>
using namespace std;  // If you omit this, then you have to refer to classes imported from std as std::, e.g., std::cout, std:array etc.

///====================================== CONTAINER ADAPTERS ================================================
/// Container adapters are stack, queue (fifo-queue), and priority_queue (binary heap) 
///
void AdapterExamples() {
#if 0
	///--------------------------------- stack ----------------------------------------
	cout << "Creating a lilo stack..." << endl;

	stack<int> S;

	cout << "push(4)" << endl;
	S.push(4);

	cout << "push(3)" << endl;
	S.push(3);

	cout << "push(7)" << endl;
	S.push(7);

	cout << "push(1)" << endl;
	S.push(1);

	cout << "push(9)" << endl;
	S.push(9);
	cout << "Number of elements in the stack: " << S.size() << endl;
	cout << "Element at the top of the stack: " << S.top() << endl;
	cout << "Removing the top element of the stack." << endl;
	S.pop();
	cout << "Element at the top of the stack now: " << S.top() << endl;
	cout << "Is the stack empty? " << (S.empty()? "Yes" : "No") << endl;

	cout << "Removing the stack elements in order: ";
	while (!S.empty()) {
		cout << S.top() << ", ";
		S.pop();
	} //end-while
	cout << endl;
	cout << "Is the stack empty? " << (S.empty() ? "Yes" : "No") << endl;
	cout << "----------------------------------------------------------" << endl;
#endif

#if 0
	///------------------------ queue ---------------------------------
	cout << "Creating a fifo queue..." << endl;

	queue<int> Q;

	cout << "Enqueue(4)" << endl;
	Q.push(4);

	cout << "Enqueue(3)" << endl;
	Q.push(3);

	cout << "Enqueue(7)" << endl;
	Q.push(7);

	cout << "Enqueue(1)" << endl;
	Q.push(1);

	cout << "Enqueue(9)" << endl;
	Q.push(9);
	cout << "Number of elements in the queue: " << Q.size() << endl;
	cout << "Element at the front of the queue: " << Q.front() << endl;
	cout << "Removing the front element of the queue." << endl;
	Q.pop();
	cout << "Element at the front of the queue now: " << Q.front() << endl;
	cout << "Is the queue empty? " << (Q.empty() ? "Yes" : "No") << endl;

	cout << "Removing the queue elements in order: ";
	while (!Q.empty()) {
		cout << Q.front() << ", ";
		Q.pop();
	} //end-while
	cout << endl;
	cout << "Is the queue empty? " << (Q.empty() ? "Yes" : "No") << endl;
	cout << "----------------------------------------------------------" << endl;
#endif

#if 0
	/// Deque example
	cout << "Creating a deque..." << endl;
	deque<int> dq;

	cout << "push_back(4)" << endl;
	dq.push_back(4);

	cout << "push_back(7)" << endl;
	dq.push_back(7);

	cout << "push_front(9)" << endl;
	dq.push_front(9);

	cout << "push_front(11)" << endl;
	dq.push_front(11);

	cout << "Number of elements in the deque: " << dq.size() << endl;
	cout << "Element at the front of the deque: " << dq.front() << endl;
	cout << "Element at the end of the deque: " << dq.back() << endl;
	cout << "Removing the front element of the deque." << endl;
	dq.pop_front();
	cout << "Element at the front of the deque now: " << dq.front() << endl;

	cout << "Removing the last element of the deque." << endl;
	dq.pop_back();
	cout << "Element at the front of the deque now: " << dq.back() << endl;

	cout << "Is the deque empty? " << (dq.empty() ? "Yes" : "No") << endl;

	cout << "Removing the deque elements from the front end: ";
	while (!dq.empty()) {
		cout << dq.front() << ", ";
		dq.pop_front();
	} //end-while
	cout << endl;
	cout << "Is the deque empty? " << (dq.empty() ? "Yes" : "No") << endl;
	cout << "----------------------------------------------------------" << endl;
#endif

#if 0
	///------------------------ priority queue ---------------------------------
	cout << "Creating a max-heap..." << endl;
	priority_queue<int> H;

	cout << "Insert(4)" << endl;
	H.push(4);

	cout << "Insert(3)" << endl;
	H.push(3);

	cout << "Insert(7)" << endl;
	H.push(7);

	cout << "Insert(1)" << endl;
	H.push(1);

	cout << "Insert(9)" << endl;
	H.push(9);
	cout << "Number of elements in the heap: " << H.size() << endl;
	cout << "Element at the top of the heap: " << H.top() << endl;
	cout << "Removing the top element of the heap." << endl;
	H.pop();
	cout << "Element at the top of the heap now: " << H.top() << endl;
	cout << "Is the heap empty? " << (H.empty() ? "Yes" : "No") << endl;

	cout << "Removing the heap elements in order: ";
	while (!H.empty()) {
		cout << H.top() << ", ";
		H.pop();
	} //end-while
	cout << endl;
	cout << "Is the heap empty? " << (H.empty() ? "Yes" : "No") << endl;
	cout << "----------------------------------------------------------" << endl;
#endif

#if 1
	///----------------------------------------------------------
	cout << "Now creating a min-heap" << endl;
#if 1
	auto cmp = [](int left, int right) { return (left >= right); };
	std::priority_queue<int, std::vector<int>, decltype(cmp)> H2(cmp);
#else
	// Alternatively, you can use standard template classes less & greater to create a max/min heap
	// This is only possible after C++17
	std::priority_queue<int, std::vector<int>, std::greater<int>> H2;
#endif

	cout << "Insert(4)" << endl;
	H2.push(4);

	cout << "Insert(3)" << endl;
	H2.push(3);

	cout << "Insert(7)" << endl;
	H2.push(7);

	cout << "Insert(1)" << endl;
	H2.push(1);

	cout << "Insert(9)" << endl;
	H2.push(9);
	cout << "Number of elements in the heap: " << H2.size() << endl;
	cout << "Element at the top of the heap: " << H2.top() << endl;
	cout << "Removing the top element of the heap." << endl;
	H2.pop();
	cout << "Element at the top of the heap now: " << H2.top() << endl;
	cout << "Is the heap empty? " << (H2.empty() ? "Yes" : "No") << endl;

	cout << "Removing the heap elements in order: ";
	while (!H2.empty()) {
		cout << H2.top() << ", ";
		H2.pop();
	} //end-while
	cout << endl;
	cout << "Is the heap empty? " << (H2.empty() ? "Yes" : "No") << endl;
	cout << "----------------------------------------------------------" << endl;
#endif
} //end-AdapterExamples

File: ADTs/Java-DS/AdapterExamples.java
///------------------------------------------------------------------------------------
/// For further reference, see: https://www.javatpoint.com/collections-in-java
///------------------------------------------------------------------------------------
/// Broadly speaking, the containers in Java are divided into 4 broad categories:
/// 1. Sequence Containers (a.k.a. lists): ArrayList, LinkedList
/// 2. Associative Containers (a.k.a. balanced trees): TreeSet [Stores just the key values], TreeMap [Stores <key, value> pairs]
/// 3. Unordered Associative Containers (a.k.a. hash tables): HashSet, HashMap
/// 4. Container Adapters: Stack, ArrayDeque, PriorityQueue (a.k.a., binary heap)

import java.util.*;

class AdapterExamples {
	static public void main(String argv[]){
  	    AdapterExamples();   // Stack, Queue, PriorityQueue
	} //end-main

	///====================================== CONTAINER ADAPTERS ================================================
	/// Container adapters are stack, queue (fifo-queue), and priority_queue (binary heap)
	///
	public static void AdapterExamples() {
		///--------------------------------- Stack ----------------------------------------
		System.out.println("Creating a lilo Stack...");

		Stack<Integer> S = new Stack<Integer>();

		System.out.println("push(4)");
		S.push(4);

		System.out.println("push(3)");
		S.push(3);

		System.out.println("push(7)");
		S.push(7);

		System.out.println("push(1)");
		S.push(1);

		System.out.println("push(9)");
		S.push(9);

		System.out.println("Number of elements in the stack: " + S.size());
		System.out.println("Element at the top of the stack: " + S.peek());
		System.out.println("Removing the top element of the stack.");
		System.out.println("Element at the top of the stack now: " + S.pop());
		System.out.println("Is the stack empty? " + (S.empty()? "Yes" : "No"));

		System.out.println("Removing the stack elements in order: ");
		while (!S.empty()) {
			System.out.print(S.pop() + ", ");
		} //end-while
		System.out.println();
		System.out.println("Is the stack empty? " + (S.empty() ? "Yes" : "No"));
		System.out.println("----------------------------------------------------------");

		///------------------------ queue ---------------------------------
		System.out.println("Creating a fifo queue...");

		Queue<Integer> Q = new LinkedList<>();

		System.out.println("Enqueue(4)");
		Q.add(4);

		System.out.println("Enqueue(3)");
		Q.add(3);

		System.out.println("Enqueue(7)");
		Q.add(7);

		System.out.println("Enqueue(1)");
		Q.add(1);

		System.out.println("Enqueue(9)");
		Q.add(9);
		System.out.println("Number of elements in the queue: " + Q.size());
		System.out.println("Element at the front of the queue: " + Q.peek());
		System.out.println("Removing the front element of the queue.");
		System.out.println("Element at the front of the queue now: " + Q.remove());
		System.out.println("Is the queue empty? " + (Q.isEmpty() ? "Yes" : "No"));

		System.out.print("Removing the queue elements in order: ");
		while (!Q.isEmpty()) {
			System.out.print(Q.remove() + ", ");
		} //end-while
		System.out.println();
		System.out.println("Is the queue empty? " + (Q.isEmpty() ? "Yes" : "No"));
		System.out.println("----------------------------------------------------------");

		///------------------------ deque ---------------------------------
		System.out.println("Creating a deque...");
		Deque<Integer> dq = new LinkedList<Integer>();

		System.out.println("addLast(4)");
		dq.addLast(4);

		System.out.println("addLast(7)");
		dq.addLast(7);

		System.out.println("addFirst(9)");
		dq.addFirst(9);

		System.out.println("addFirst(11)");
		dq.addFirst(11);

		System.out.println("Number of elements in the deque: " + dq.size());
		System.out.println("Element at the front of the deque: " + dq.peekFirst());
		System.out.println("Element at the end of the deque: " + dq.peekLast());
		System.out.println("Removing the front element of the deque.");
		dq.removeFirst();
		System.out.println("Element at the front of the deque now: " + dq.peekFirst());

		System.out.println("Removing the last element of the deque.");
		dq.removeLast();
		System.out.println("Element at the front of the deque now: " + dq.peekLast());		
		System.out.println("Is the deque empty? " + (dq.isEmpty() ? "Yes" : "No"));

		System.out.print("Removing the deque elements from the front end: ");
		while (!dq.isEmpty()) {
			System.out.print(dq.removeFirst() + ", ");
		} //end-while
		System.out.println();
		System.out.println("Is the deque empty? " + (dq.isEmpty() ? "Yes" : "No"));
		System.out.println("----------------------------------------------------------");			

		///------------------------ priority queue ---------------------------------
		System.out.println("Creating a min-heap...");
		PriorityQueue<Integer> H = new PriorityQueue<>();

		System.out.println("Insert(4)");
		H.add(4);

		System.out.println("Insert(3)");
		H.add(3);

		System.out.println("Insert(7)");
		H.add(7);

		System.out.println("Insert(1)");
		H.add(1);

		System.out.println("Insert(9)");
		H.add(9);
		System.out.println("Number of elements in the heap: " + H.size());
		System.out.println("Element at the top of the heap: " + H.peek());
		System.out.println("Removing the top element of the heap.");
		System.out.println("Element at the top of the heap now: " + H.poll());
		System.out.println("Is the heap empty? " + (H.size() == 0 ? "Yes" : "No"));

		System.out.print("Removing the heap elements in order: ");
		while (H.size() > 0) {
			System.out.print(H.poll() + ", ");
		} //end-while
		System.out.println();
		System.out.println("Is the heap empty? " + (H.size() == 0 ? "Yes" : "No"));
		System.out.println("----------------------------------------------------------");

		///----------------------------------------------------------
		System.out.println("Creating a max-heap...");
		PriorityQueue<Integer> H2 = new PriorityQueue<>((x, y) -> Integer.compare(y, x)); // Lambda comparator

		System.out.println("Insert(4)");
		H2.add(4);

		System.out.println("Insert(3)");
		H2.add(3);

		System.out.println("Insert(7)");
		H2.add(7);

		System.out.println("Insert(1)");
		H2.add(1);

		System.out.println("Insert(9)");
		H2.add(9);
		System.out.println("Number of elements in the heap: " + H2.size());
		System.out.println("Element at the top of the heap: " + H2.peek());
		System.out.println("Removing the top element of the heap.");
		System.out.println("Element at the top of the heap now: " + H2.poll());
		System.out.println("Is the heap empty? " + (H2.size() == 0 ? "Yes" : "No"));

		System.out.print("Removing the heap elements in order: ");
		while (H2.size() > 0) {
			System.out.print(H2.poll() + ", ");
		} //end-while
		System.out.println();
		System.out.println("Is the heap empty? " + (H2.size() == 0 ? "Yes" : "No"));
		System.out.println("----------------------------------------------------------");
	} //end-AdapterExamples

};